pkg/binlog/event/event.go

```diff
@@ -0,0 +1,759 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// binlog events generator for MySQL used to generate some binlog events for tests.
+// Readability takes precedence over performance.
+
+package event
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/pingcap/errors"
+	gmysql "github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+
+	"github.com/pingcap/dm/pkg/gtid"
+)
+
+// flags used in RowsEvent.
+const (
+	RowFlagsEndOfStatement     uint16 = 0x0001
+	RowFlagsNoForeignKeyChecks uint16 = 0x0002
+	RowFlagsNoUniqueKeyChecks  uint16 = 0x0004
+	RowFlagsRowHasAColumns     uint16 = 0x0008
+)
+
+const (
+	// GTIDFlagsCommitYes represents a GTID flag with [commit=yes].
+	// in `Row` binlog format, this will appear in GTID event before DDL query event.
+	GTIDFlagsCommitYes uint8 = 1
+
+	binlogVersion   uint16 = 4 // only binlog-version 4 supported now
+	mysqlVersion           = "5.7.22-log"
+	mysqlVersionLen        = 50                                 // fix-length
+	eventHeaderLen         = uint8(replication.EventHeaderSize) // always 19
+	crc32Len        uint32 = 4                                  // CRC32-length
+	tableMapFlags   uint16 = 1                                  // flags in TableMapEvent's post-header, not used yet
+)
+
+var (
+	// A array indexed by `Binlog-Event-Type - 1` to extract the length of the event specific header.
+	// It is copied from a binlog file generated by MySQL 5.7.22-log.
+	// The doc at https://dev.mysql.com/doc/internals/en/format-description-event.html does not include all of them.
+	eventTypeHeaderLen = []byte{0x38, 0x0d, 0x00, 0x08, 0x00, 0x12, 0x00, 0x04, 0x04, 0x04, 0x04, 0x12, 0x00, 0x00, 0x5f, 0x00, 0x04, 0x1a, 0x08, 0x00, 0x00, 0x00, 0x08, 0x08, 0x08, 0x02, 0x00, 0x00, 0x00, 0x0a, 0x0a, 0x0a, 0x2a, 0x2a, 0x00, 0x12, 0x34, 0x00}
+)
+
+// GenEventHeader generates a EventHeader's raw data according to a passed-in EventHeader struct.
+// ref: https://dev.mysql.com/doc/internals/en/binlog-event-header.html
+func GenEventHeader(header *replication.EventHeader) ([]byte, error) {
+	buf := new(bytes.Buffer)
+
+	// timestamp, 4 bytes
+	err := binary.Write(buf, binary.LittleEndian, header.Timestamp)
+	if err != nil {
+		return nil, errors.Annotatef(err, "write timestamp %d", header.Timestamp)
```

Comment:
do you want add "[generate event header]" information in the error?

Reply:
I tend to:

if `A` call `B`, `B` call `C`, and `C` return an error.
then
1. annotate `err` in `B` with something like `calling C fail`
2. annotate `err` in `A` with something like `calling B fail`

use this way, if `B` also call `D` and `D` return an error, then only need annotate `err` with something like `calling D` fail.

so, in this L66, I only annotate the error with information about L64, and annotate something like `generate event header` in the outer function.

Reply:
ok. Just be unified